### PR TITLE
Feature/additional client configuration

### PR DESCRIPTION
**Pull request by ivanvecko**

### CODE REVIEW

File: Versioner/PrinceOfVersions.swift [hunk 1]
```diff
@@ -103,5 +119,69 @@ public struct PrinceOfVersions {
...
+            if hasKey {
+                completionHandler(.useCredential, URLCredential(trust: trust))
+                return
+            }
+        }
+        
+        completionHandler(.cancelAuthenticationChallenge, nil)
+    }
+    
+    fileprivate func pinnedKeys() -> [SecKey] {
```
Comment (damjanvujaklija) [inline Versioner/PrinceOfVersions.swift]: Private methods should have underscore prefix.

**Comment by damjanvujaklija** [comment]
Added one minor comment but other than that seems ok. Though I didn't do a lot of apps with certificate pinning so best someone who has this experience also goes through the code.

File: Versioner/PrinceOfVersions.swift (hunk 1 above)
Comment (Adis) [inline Versioner/PrinceOfVersions.swift]: Ako dohvacate cert koji se pina online, onda nema smisla pinnat cert, buduci da nije dosao sigurnim kanalom. Cert mora ili biti vec dodan u app, ili pinnanje nema smisla, tako da mi ova metoda ne radi nista posebno. Full je lagano podvaliti fejk cert, buduci da cemo ga ionako skinuti pa kao pinnati.

Comment (ivanvecko) [reply Versioner/PrinceOfVersions.swift]: Hm, nisam te bas shvation na sto mislis. Ne dohvaca se cert online pa se onda postavlja za pinnanje. URL koji se predaje je lokalni. Znaci nesto ovako: `Bundle.main.url(forResource: "certificate", withExtension: "crt")`

Comment (Adis) [reply Versioner/PrinceOfVersions.swift]: Ohhh. Bilo mi je cudno jer izgleda kao da ga dohvacas remotely, sto je isto teoretski moguce. Jel PoV pod ili? Jer mozes procitat sve certove iz nekog bundlea, tako to radi Alamofire.

Comment (nsimunko) [reply Versioner/PrinceOfVersions.swift]: Fixed.

Comment (nsimunko) [reply Versioner/PrinceOfVersions.swift]: Implementirano iteriranje kroz sve certifikate u main bundle i pinnanje svih keyeva (kao kod Alamofirea). Also, updetao sam README file sa sampleom za SecCert usage.

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+version: 2
```
Comment (Truba) [inline .circleci/config.yml]: We can move it to Bitrise as well.

File: PrinceOfVersions.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -440,7 +456,7 @@
 				MTL_ENABLE_DEBUG_INFO = NO;
 				SDKROOT = iphoneos;
 				SWIFT_OPTIMIZATION_LEVEL = "-Owholemodule";
-				SWIFT_VERSION = 3.0;
+				SWIFT_VERSION = 4.2;
```
Comment (Truba) [inline PrinceOfVersions.xcodeproj/project.pbxproj]: Also please add version to podfile.

File: Versioner/PrinceOfVersions.swift [hunk 4]
```diff
@@ -29,15 +29,21 @@ public struct PrinceOfVersions {
     public typealias NoNewVersionBlock = (Bool, [String: Any]?) -> Void
     public typealias ErrorBlock = (Error) -> Void
 
-    public init(){}
+    fileprivate var _shouldPinCertificates: Bool = false
```
Comment (Truba) [inline Versioner/PrinceOfVersions.swift]: Prefer `private` here.

File: Versioner/PrinceOfVersions.swift [hunk 5]
```diff
@@ -52,10 +58,20 @@ public struct PrinceOfVersions {
...
     @discardableResult
-    public func loadConfiguration(from URL: URL, completion: @escaping CompletionBlock) -> URLSessionDataTask {
-
-        let defaultSession = URLSession(configuration: URLSessionConfiguration.default)
-        let dataTask = defaultSession.dataTask(with: URL, completionHandler: { (data, response, error) in
+    public func loadConfiguration(from URL: URL, httpHeaderFields: [String : String?]? = nil, shouldPinCertificates: Bool = false, completion: @escaping CompletionBlock) -> URLSessionDataTask {
+
+        _shouldPinCertificates = shouldPinCertificates
+        
+        let defaultSession = URLSession(configuration: URLSessionConfiguration.default, delegate: self, delegateQueue: nil)
```
Comment (Truba) [inline Versioner/PrinceOfVersions.swift]: Can you also please expose `delegateQueue`

File: Versioner/PrinceOfVersions.swift [hunk 6]
```diff
@@ -103,5 +121,83 @@ public struct PrinceOfVersions {
...
+            if hasKey {
+                completionHandler(.useCredential, URLCredential(trust: trust))
+                return
+            }
+        }
+        
+        completionHandler(.cancelAuthenticationChallenge, nil)
+    }
+    
+    fileprivate func _certificates(in bundle: Bundle = Bundle.main) -> [SecCertificate] {
```
Comment (Truba) [inline Versioner/PrinceOfVersions.swift]: private

File: Versioner/PrinceOfVersions.swift [hunk 7]
```diff
@@ -103,5 +121,83 @@ public struct PrinceOfVersions {
...
+                let certificate = SecCertificateCreateWithData(nil, certificateData)
+            {
+                certificates.append(certificate)
+            }
+        }
+        
+        return certificates
+    }
+    
+    fileprivate func _pinnedKeys() -> [SecKey] {
```
Comment (Truba) [inline Versioner/PrinceOfVersions.swift]: private

File: Versioner/PrinceOfVersions.swift [hunk 8]
```diff
@@ -103,5 +121,83 @@ public struct PrinceOfVersions {
...
+        for certificate in _certificates() {
+            if let key = _publicKey(for: certificate) {
+                publicKeys.append(key)
+            }
+        }
+        
+        return publicKeys
+    }
+    
+    fileprivate func _publicKey(for certificate: SecCertificate) -> SecKey? {
```
Comment (Truba) [inline Versioner/PrinceOfVersions.swift]: private

File: Versioner/PrinceOfVersions.swift [hunk 9]
```diff
@@ -103,5 +127,83 @@ public struct PrinceOfVersions {
...
+        }
+        
+        if let serverCertificate = SecTrustGetCertificateAtIndex(trust, 0), let serverCertificateKey = _publicKey(for: serverCertificate) {
+            
+            let hasKey = _pinnedKeys().contains(where: { (key) -> Bool in
+                if (serverCertificateKey as AnyObject).isEqual(key) {
+                    return true
+                } else {
+                    return false
+                }
```
Comment (jabou) [inline Versioner/PrinceOfVersions.swift]: why not just: `return (serverCertificateKey as AnyObject).isEqual(key)`?

**Pull request merged**